### PR TITLE
Docs: Make docsite rebuilds smarter/faster

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
The purpose of this PR is to make rebuilding the docsite as fast as possible by ensuring that we only need to rebuild what has changed, rather than rebuild everything all the time.

This PR includes:
- Disable automatic cleanup of generated files (`make clean` becomes explicit)
  - Processed files go **from 2502 to 2092** *(rebuilding the user documentation)*
- Make `plugin_formatter` idempotent (rewrite file only if it has changed)
  - Processed files go **from 2092 to 12** *(rebuilding modules + plugins docs)*
- Make `generate_man.py` idempotent
  - Processed files go **from 12 to 3**
- Make `dump_keywords.py` idempotent
  - Processed files go **from 3 to 2**
- Make `dump_config.py` idempotent
  - Processed files go **from 2 to 1**
- Make `testing_formatter.sh` idempotent
  - Processed files go **from 1 to 0**

PS This means that for e.g. rebuilding the docs for the website, you probably want to run `make clean` explicitly first.

##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
docsite

##### ANSIBLE VERSION
v2.8

### CODE REVIEW

**Comment by dagwieers** [comment]
cc @felixfontein

File: docs/bin/dump_config.py
Comment (felixfontein) [inline docs/bin/dump_config.py]: Why not simply compare contents, instead of computing sha1 of both first? Since you read both into memory, that's more efficient :)

Also: it currently fails under Python 3:
```
Traceback (most recent call last):
  File "../bin/dump_config.py", line 84, in <module>
    sys.exit(main(sys.argv[:]))
  File "../bin/dump_config.py", line 72, in main
    sha1_old = sha1(open(output_name).read()).hexdigest()
TypeError: Unicode-objects must be encoded before hashing
```

Comment (felixfontein) [reply docs/bin/dump_config.py]: Also: since you're not closing the file, this might break under strange file systems (such under Windows -- no idea if anyone ever tries to do that, tough :) ).

Comment (dagwieers) [reply docs/bin/dump_config.py]: @felixfontein When the script finishes, the file is closed. I don't think it is a problem as every call generates a single file anyway.

Comment (felixfontein) [reply docs/bin/dump_config.py]: It probably makes sense to create a new `utils.py` module which has a function `write_to_file_if_different(filename, data)` which does the heavy lifting (and where this is done cleanly), so that all the details are only repeated once.

**Comment by felixfontein** [comment]
For me, with this change (adjusted to work with Python 3 and doing direct comparsions instead of computing `sha1`), I got some timing improvements when running the update docs process for six modules. First run:
```
real	2m18.633s
user	6m0.063s
sys	0m4.443s
```
Running it again (with no change):
```
real	1m10.700s
user	2m40.249s
sys	0m2.763s
```
So yeah, my laptop still gets warm (and I can hear when it's done when the fan stops making noise), but it got much better :)

**Comment by dagwieers** [comment]
What command did you use ? I can run it under 12 seconds when there are no changes.

    real	0m11.330s
    user	0m10.443s
    sys	0m0.889s


*(Doing sha1 checksums or comparing file content doesn't make a real impact)*

File: docs/bin/dump_keywords.py
Comment (felixfontein) [inline docs/bin/dump_keywords.py]: You need to `keyword_page.encode('utf-8')` here (before, the file was opened in text mode).

File: docs/bin/plugin_formatter.py
Comment (felixfontein) [inline docs/bin/plugin_formatter.py]: You need either `to_bytes(text)` or `text.encode('utf-8')` here as well.

**Review by felixfontein — CHANGES_REQUESTED** [review]
Some more changes for Python 3 :)

**Comment by felixfontein** [comment]
I guess I screwed something up in my corrections yesterday; with the newest version (plus the two changes for Python 3), it's now down to 11 seconds if nothing changed.

**Comment by gundalow** [comment]
One issue maybe if someone locally is building docs, swap branches then builds docs again.

This can partially be fixed by backporting all these docs changes to the current `stable-2.` branches

File: docs/bin/dump_config.py
Comment (webknjaz) [inline docs/bin/dump_config.py]: Please keep 0 return code. This is a way of communicating success to an OS

Comment (dagwieers) [reply docs/bin/dump_config.py]: Which is the default, so it's useless.

Comment (dagwieers) [reply docs/bin/dump_config.py]: Besides, `return` here is what the `main()` function returns, not the program's exit code (which would be using `sys.exit(0)`)

File: docs/bin/dump_keywords.py
Comment (dagwieers) [reply docs/bin/dump_keywords.py]: Fixed.

File: docs/bin/plugin_formatter.py
Comment (dagwieers) [reply docs/bin/plugin_formatter.py]: Fixed.

**Comment by felixfontein** [comment]
@gundalow That shouldn't be a problem (except maybe that `make clean` isn't executed): if you're on a `stable-xxx` branch, all files will be rebuilt in any case, and if you switch back to `devel`, the extractors will compare all `.rst` files and -- if needed -- overwrite them all. So nothing should go wrong. (Except, as said, the missing `make clean`, which could have effects on doc building.)

**Comment by dagwieers** [comment]
@gundalow The other branches still do `make clean` if you do `make webdocs`, so I think that's fine.

File: docs/bin/dump_keywords.py
Comment (felixfontein) [inline docs/bin/dump_keywords.py]: This isn't complete: for Python 3 (with a new enough Jinja2), it still breaks as `to_bytes` isn't called. Can you move the `to_bytes` to the `update_file_if_different()` call, like this?
```
update_file_if_different(outputname, to_bytes(keyword_page))
```

File: docs/bin/dump_config.py
Comment (felixfontein) [reply docs/bin/dump_config.py]: The place where the `main()` function is called explicitly does `sys.exit(main(sys.argv[:]))`. So `return 0` should really be here.

(And yes, returning nothing is equivalent to `return None`, and `sys.exit(None)` is equivalent to `sys.exit(0)`, but it's still nicer to have an explicit `return 0` IMO.)

**Review by felixfontein — CHANGES_REQUESTED** [review]

File: docs/bin/dump_config.py
Comment (dagwieers) [reply docs/bin/dump_config.py]: There's nothing in that script that returns anything else though, so to me it's only adding confusion. The perception that something greater is going on, but there's nothing behind the curtain.

Comment (felixfontein) [reply docs/bin/dump_config.py]: In that case, you should also remove the `sys.exit()` call around `main(sys.argv[:])` (in `docs/bin/plugin_formatter.py`, there's also no `sys.exit()` around `main()`).

Comment (dagwieers) [reply docs/bin/dump_config.py]: I did that when you made me aware of it :-) Thanks BTW !

**Review by felixfontein — APPROVED** [review]
Works fine with Python 3 now. And 11 seconds definitely beats > 2 minutes!

**Review by achinthagunasekara — APPROVED** [review]

File: docs/bin/dump_config.py
Comment (webknjaz) [inline docs/bin/dump_config.py]: Passing `sys.argv[:]` here is useless wrt current implementation.
If you look at the `main()`'s first three lines there's:
```python
def main(args):
    ...
    (options, args) = parser.parse_args()  # <-- immediatelly rewrites args variable
```

It looks like the initial implementation has been written by the person with some C-like background, where main accepts args data and returns 0. But in fact in this case it's not needed.

Please remove this arg.

Comment (dagwieers) [reply docs/bin/dump_config.py]: @webknjaz You are correct, and this code, as well as the other scripts are not beauties. But I do not intend to rewrite complete scripts. Can you maybe propose this in a subsequent PR once this one is merged ? A cleanup would be very welcome indeed.

Comment (webknjaz) [reply docs/bin/dump_config.py]: Oops, I missed your comment and already did that in commit...

Comment (webknjaz) [reply docs/bin/dump_config.py]: Anyway, removal of `return 0` was not related to the purpose of the PR, but related to my refactoring suggestion. So it fits next to your change :)

Comment (dagwieers) [reply docs/bin/dump_config.py]: @webknjaz Yeah, but if you continue in that line, you end up rewriting all the scripts and that is not something your colleagues would accept anyway.

Comment (webknjaz) [reply docs/bin/dump_config.py]: Agreed

**Comment by dagwieers** [comment]
@acozine Can we please get this merged ? I am working on documentation again, and not having to rebuild (almost) everything makes a real difference. I don't want to wait one hour to rebuild one single document to see if my statement was correct.

File: docs/bin/dump_config.py
Comment (abadger) [inline docs/bin/dump_config.py]: ```return 0``` and ```sys.exit(main(sys.argv[:]))``` should remain.

Keeping those allows the script to return a different exit code on error and isn't unusual or confusing for someone who is used to UNIX programming.

File: docs/bin/dump_config.py
Comment (abadger) [reply docs/bin/dump_config.py]: I agree with dag that we don't want to change args handling in this PR but for the record the correct change here is to use args in the call to parse_args rather than removing sys.argv from the call to main():

```
def main(args):
    ...
    (options, args) = parser.parse_args(args)
[...]

if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
```

Code that follows this style is easier to test and easier to reason about.

File: lib/ansible/utils/helpers.py
Comment (abadger) [inline lib/ansible/utils/helpers.py]: This doesn't really belong here as nothing in ansible's runtime needs it.  Unfortunately, we don't have a directory where it would be picked up by the build scripts.  So maybe a separate, private file in here like lib/ansible/utils/_build_helpers.py.

**Comment by abadger** [comment]
The logic here looks good.  A couple of notes to make the code better (and isolate it so that it becomes easier for other RMs to approve potential backports of this).

**Comment by abadger** [comment]
Regarding the swapping branches case... there will be a cornercase problem where generated rst files that are not present in the other branch are now not cleaned before the branch is built (This also affects devel when a module or other documentation is removed).  Using Makefiles usually works around this by having a master list of all source files somewhere in the Makefile but we don't have that in the docs build.

Note that this is not a new problem.  When module docs moved to a subdirectory I found that I had a lot of crufty files left over in the directory.  I would have had to make clean prior to checking out the module docs move in order to clean those up.

File: docs/bin/dump_config.py
Comment (dagwieers) [reply docs/bin/dump_config.py]: Well, none of the other scripts are doing this. And it adds absolutely no value as this script was not designed to return anything else whatsoever. So this is nothing more than a mental exercise.

So you want me to add this to the other scripts ? Do we also add stuff like this to modules ? I hope we don't.

**Comment by dagwieers** [comment]
@abadger Sure, as I stated, doing a `make clean` becomes a conscious decision rather a mandatory action. Not making this possible is simply not acceptable if you want people to enjoy improving documentation (whether it means a module doc or a small change in a guide).

I don't mind creating a separate target in the Makefile that doesn't do the clean, but IMO that muddies everything even more. The only situation where this is a real problem is the docs generated for the website (which is a contained process we manage), for most contributors having stale, unused, unreferenced leftovers is not a real concern IMO. And if so, we simply have to document `make clean` in the process.

PS I don't think we want to backport this. Documentation work is mostly done (and has the most impact) in the devel branch.

File: lib/ansible/utils/doc_helpers.py
Comment (abadger) [inline lib/ansible/utils/doc_helpers.py]: Can we make this private (prepend the filename with a leading underscore)?  The reason is that it is only used at buildtime by our build scripts.  It isn't used at runtime at all.  Marking it private clues other people in to the fact that they shouldn't use it in runtime code.

Comment (dagwieers) [reply lib/ansible/utils/doc_helpers.py]: Well, I would prefer that if this functionality *is* useful, it does get used in runtime code, rather than being reinvented. That's why I would prefer it in a normal library, helper.py was a perfect fit IMO. But announcing it as "not to be used" seems to be the wrong reflex ?

Comment (dagwieers) [reply lib/ansible/utils/doc_helpers.py]: There's more code that is being used at build-time coming from the normal libraries.

Comment (abadger) [reply lib/ansible/utils/doc_helpers.py]: Do you have a use  case in the runtime code in mind that you can point me at?

Comment (abadger) [reply lib/ansible/utils/doc_helpers.py]: and yes, build time using run time code is fine.  What's misorganized is when build time uses runtime code which nothing in runtime uses.

File: docs/bin/dump_config.py
Comment (abadger) [reply docs/bin/dump_config.py]: We could add it to other scripts.

modules have their own exit points (fail_json and exit_json) which already call sys.exit(1) and sys.exit(0) respectively.

Comment (abadger) [reply docs/bin/dump_config.py]: Note though, I wouldn't add to other scripts in this PR.... it's a separate code cleanup from the main purpose of this one (to stop rebuilding already-built docs)

**Comment by abadger** [comment]
> @abadger Sure, as I stated, doing a `make clean` becomes a conscious decision rather a mandatory action. Not making this possible is simply not acceptable if you want people to enjoy improving documentation (whether it means a module doc or a small change in a guide).
> 
> I don't mind creating a separate target in the Makefile that doesn't do the clean, but IMO that muddies everything even more.

Yeah, I agree... don't create a separate Makefile target.

My note isn't really about running make clean... it's more that make clean is insufficient and we're slightly increasing the boundaries where that is the case.  We can fix make clean by specifying all of the source files in the Makefile or we can continue to ignore those cornercases.  Since it's an existing problem, docsite rebuild should be designed to use a clean checkout rather than relying on make clean.

File: docs/bin/dump_config.py
Comment (dagwieers) [reply docs/bin/dump_config.py]: No, I think it's useless but leave it up to @webknjaz to rewrite those scripts as he clearly sees great benefits :-)

File: lib/ansible/utils/doc_helpers.py
Comment (dagwieers) [reply lib/ansible/utils/doc_helpers.py]: Private normally means 'not exposed for external use', because you want to avoid people to use it. I don't see why this function should not be used. This is another mental exercise and very ambiguous. Should we avoid people using *this* function, is there a greater harm that I am not seeing or should we encourage to reuse it if it is useful ? And why make an exceptional case for one simple function.

Currently we have utils/plugins_docs.py which is not private either, contains 2 functions that are only used by the same scripts. So I think you all have too much time on your hands to split hairs for things that don't matter as much as you think they do. Maybe, you can do those cleanups in a separate PR and merge this. Because I have had it with those small pet-peeves. I'd like to spend my time on more important things, if you don't mind :-)

Comment (abadger) [reply lib/ansible/utils/doc_helpers.py]: plugin_docs is an example of something that should not have been committed into utils without more thought.... however, if you ore closely, get_docstring() is in multiple files in runtime.  it uses add_fragments.  add_fragments use merge_fragment.  So everything in utils/plugin_docs is used by runtime.

If we have a runtime use case for this, then I see value in making sure it's a good api and looking at making it good for general usage.  If we don't have a runtime usage then we end up having to maintain an api which may not be well suited to the task people put it too.  that's why things not used at runtime should remain private until a use case emerges.  Then we can make sure the API is sufficient for the use cases and make it public.

**Comment by abadger** [comment]
Updated the PR to address those concerns and merged to devel.

**Comment by dagwieers** [comment]
Thanks :-)

**Pull request merged**
